geomstats/stiefel.py

```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
+        """
+        tangent_vec_a = gs.to_ndarray(tangent_vec_a, to_ndim=3)
+        tangent_vec_b = gs.to_ndarray(tangent_vec_b, to_ndim=3)
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        base_point_transpose = gs.transpose(base_point, axes=(0, 2, 1))
+
+        aux = gs.matmul(
+            gs.transpose(tangent_vec_a, axes=(0, 2, 1)),
+            gs.eye(self.n) - 0.5 * gs.matmul(base_point, base_point_transpose))
+        inner_prod = gs.trace(gs.matmul(aux, tangent_vec_b), axis1=1, axis2=2)
+
+        inner_prod = gs.to_ndarray(inner_prod, to_ndim=2, axis=1)
+        return inner_prod
+
+    def exp(self, tangent_vec, base_point):
+        """
+        Riemannian exponential of a tangent vector wrt to a base point.
+        """
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=3)
+        n_tangent_vecs, _, _ = tangent_vec.shape
+
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        n_base_points, n, p = base_point.shape
+
+        assert (n_tangent_vecs == n_base_points
+                or n_tangent_vecs == 1
+                or n_base_points == 1)
+
+        matrix_a = gs.einsum(
+            'nij, njk->nik',
+            gs.transpose(base_point, axes=(0, 2, 1)), tangent_vec)
+        matrix_k = (tangent_vec
+                    - gs.einsum('nij,njk->nik', base_point, matrix_a))
+
+        matrix_q = gs.zeros_like(matrix_k)
+        matrix_r = gs.zeros(
+            (matrix_k.shape[0], matrix_k.shape[2], matrix_k.shape[2]))
+        for i, k in enumerate(matrix_k):
+            matrix_q[i], matrix_r[i] = gs.linalg.qr(k)
+
+        matrix_ar = gs.concatenate(
+            [matrix_a,
+             -gs.transpose(matrix_r, axes=(0, 2, 1))],
+            axis=2)
+
+        n_matrix_r = matrix_r.shape[0]
+        if n_matrix_r == 1:
+            matrix_r = gs.tile(matrix_r, (n_base_points, 1, 1))
+
+        zeros = gs.zeros((n_base_points, p, p))
+        if n_base_points == 1:
+            zeros = gs.zeros((n_tangent_vecs, p, p))
+        matrix_rz = gs.concatenate(
+            [matrix_r,
+             zeros],
+            axis=2)
+        block = gs.concatenate([matrix_ar, matrix_rz], axis=1)
+        matrix_mn_e = gs.expm(block)
+
+        if n_base_points == 1:
+            base_point = gs.tile(base_point, (n_tangent_vecs, 1, 1))
+        n_matrix_q = matrix_q.shape[0]
+        if n_matrix_q == 1:
+            matrix_q = gs.tile(matrix_q, (n_base_points, 1, 1))
+
+        exp = gs.einsum(
+            'nij,njk->nik',
+            gs.concatenate(
+                [base_point,
+                 matrix_q],
+                axis=2),
+            matrix_mn_e[:, :, 0:p])
+
+        return exp
+
+    def log(self, point, base_point, max_iter=100, tol=1e-6):
+        """
+        Riemannian logarithm of a point wrt a base point.
+
+        Based on:
+        Zimmermann, Ralf
+        "A Matrix-Algebraic Algorithm for the Riemannian Logarithm
```

Comment:
can we start adding links to the papers? would be great.

Reply:
Done.